src/rp2_common/hardware_rtc/rtc.c

```diff
@@ -114,60 +139,88 @@ void rtc_enable_alarm(void) {
-                          ((t->month < 0) ? 0 : (((uint)t->month) << RTC_IRQ_SETUP_0_MONTH_LSB)) |
-                          ((t->day   < 0) ? 0 : (((uint)t->day)   << RTC_IRQ_SETUP_0_DAY_LSB  ));
-    rtc_hw->irq_setup_1 = ((t->dotw  < 0) ? 0 : (((uint)t->dotw)  << RTC_IRQ_SETUP_1_DOTW_LSB)) |
-                          ((t->hour  < 0) ? 0 : (((uint)t->hour)  << RTC_IRQ_SETUP_1_HOUR_LSB)) |
-                          ((t->min   < 0) ? 0 : (((uint)t->min)   << RTC_IRQ_SETUP_1_MIN_LSB )) |
-                          ((t->sec   < 0) ? 0 : (((uint)t->sec)   << RTC_IRQ_SETUP_1_SEC_LSB ));
+    rtc_disable_alarm();
 
-    // Set the match enable bits for things we care about
-    if (t->year  >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_YEAR_ENA_BITS);
-    if (t->month >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_MONTH_ENA_BITS);
-    if (t->day   >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_DAY_ENA_BITS);
-    if (t->dotw  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_DOTW_ENA_BITS);
-    if (t->hour  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_HOUR_ENA_BITS);
-    if (t->min   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_MIN_ENA_BITS);
-    if (t->sec   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_SEC_ENA_BITS);
+    uint32_t s0 = 0, s1 = 0;
 
     // Does it repeat? I.e. do we not match on any of the bits
     _alarm_repeats = rtc_alarm_repeats(t);
 
+    bool check_params = (is_valid_datetime(t) || _alarm_repeats != NO_REPEAT) && user_callback;
+    valid_params_if(RTC, check_params);
+    if(!check_params)	// none of the parameters is valid
+        return false;
+
+    // Set the match enable bits for things we care about
+    if(_alarm_repeats == CONTINUOUS_REPEAT_EVERY_SEC) {
+        // repeatable every second! All entries are -1
+        datetime_t new_dt;
+        _rtc_get_datetime(&new_dt);
+        _seconds_increment = (-t->sec) % 60;
+        s1 = ADD_AND_ENABLE_REPEATABLE_SECOND(new_dt.sec);
+    }
+    else {
+        if (RANGE_CHECK_YEAR(t))  s0 |= RTC_IRQ_SETUP_0_YEAR_ENA_BITS  | (((uint)t->year)  << RTC_IRQ_SETUP_0_YEAR_LSB);
+        if (RANGE_CHECK_MONTH(t)) s0 |= RTC_IRQ_SETUP_0_MONTH_ENA_BITS | (((uint)t->month) << RTC_IRQ_SETUP_0_MONTH_LSB);
+        if (RANGE_CHECK_DAY(t))   s0 |= RTC_IRQ_SETUP_0_DAY_ENA_BITS   | (((uint)t->day)   << RTC_IRQ_SETUP_0_DAY_LSB);
+        if (RANGE_CHECK_DOTW(t))  s1 |= RTC_IRQ_SETUP_1_DOTW_ENA_BITS  | (((uint)t->dotw)  << RTC_IRQ_SETUP_1_DOTW_LSB);
+        if (RANGE_CHECK_HOUR(t))  s1 |= RTC_IRQ_SETUP_1_HOUR_ENA_BITS  | (((uint)t->hour)  << RTC_IRQ_SETUP_1_HOUR_LSB);
+        if (RANGE_CHECK_MIN(t))   s1 |= RTC_IRQ_SETUP_1_MIN_ENA_BITS   | (((uint)t->min)   << RTC_IRQ_SETUP_1_MIN_LSB);
+        if (RANGE_CHECK_SEC(t))   s1 |= RTC_IRQ_SETUP_1_SEC_ENA_BITS   | (((uint)t->sec)   << RTC_IRQ_SETUP_1_SEC_LSB);
+
+        if(!s0 && !s1) return false; // out of range datetime_t input
```

Comment:
Is this a condition that can actually happen? Won't it be caught by the earlier checks?

Reply:
it can happen if one parameter is set to a value > range and all others are -1.

Reply:
Perhaps it's worth adding an `is_valid_alarmtime` function, and using that as part of the `valid_params_if` checks?
```
static bool is_valid_alarmtime(const datetime_t *t) {
    return (RANGE_CHECK_YEAR(t) || t->year < 0)
        && (RANGE_CHECK_MONTH(t) || t->month < 0)
        && (RANGE_CHECK_DAY(t) || t->day < 0)
// etc.
```

Reply:
I don't think it's worth the effort as the function already returns false then. Just an assertion could appear - if they are enable for rtc.

I think the more critical question is, what should happen with a given user_callback, if pico is woken up from RTC event?

Reply:
> > Perhaps it's worth adding an `is_valid_alarmtime` function

> I don't think it's worth the effort

But I _think_ this would successfully reject the "a month of 14 and a day of -1" that I mentioned earlier? Meh :shrug: 

>  what should happen with a given user_callback, if pico is woken up from RTC event?

Looking at https://github.com/raspberrypi/pico-extras/blob/master/src/rp2_common/pico_sleep/sleep.c#L106 (linked to from the RP2040 datasheet) it seems that whatever callback is passed to `sleep_goto_sleep_until` (which of course might be NULL) overwrites any existing RTC user_callback.

BTW I hope you don't feel that I'm being too picky with all my comments, I'm only trying to offer constructive feedback :slightly_smiling_face:   And if any of my comments conflict with what @kilograham says, please ignore my comments and listen to Graham :grinning:

Reply:
It's fine as your comments make sense. And now, good find! Yes, at least `sleep_goto_sleep_until` tries to overwrite the user callback. But the user callback is set with `irq_set_exclusive_handler`. And now we find ourselves within an `hard_assert(current == __unhandled_user_irq || current == handler);`

So my freshly implemented rtc_delete_alarm() is directly required for  `sleep_goto_sleep_until` :-)

Reply:
> And now we find ourselves within an hard_assert(current == __unhandled_user_irq || current == handler);

Are you sure? I think you might be mixing up the irq_handler and the user_callback again?  AFAICT (and I could easily be entirely wrong!) the `sleep_goto_sleep_until` only changes the user_callback (which is called by the irq_handler), and doesn't actually change the irq_handler? :shrug:

Reply:
ok, second condition then is true (current==handler)